Sync metrics prototype

senior opened the pull request:
~~Note that the first two commits here are from https://github.com/metabase/metabase/pull/6596.~~

The primary goal of adding something like this would be to help understand what is happening with a sync process via timings and other data characteristics. As a background process, we're currently flying blind with sync. I used the timing metrics when I was trying to understand why sync was taking so long, with it I could focus my efforts are which part of the process was taking the longest.

~~This PR has some code around it and I'll attach some screenshots from JConsole of the JMX metrics.  There are other ways we can expose the data, such as JSON or build a basic UI around etc, but JConsole was easy since I can attach to the process locally. I generated an uberjar and it looked to impact the size of the jar by less than 1MB. Removed the JMX hooks as this proves to be a hurdle for users running Docker and ElasticBeanstalk.~~

This PR has morphed a bit from where it started. Originally I was adding a metrics library and leaning on that. After some discussion, it seemed like the eventual goal was to persist this data so that we can look sync runs over time (even if it was for a month or something). The metrics library I was using was just a "since startup" kind of metric. It seemed like I wasn't getting much value out of it at that point and so I ripped it out.

The bulk of the work was adding in counts to the various interesting parts of sync and analyze. Most of the functions were only side affectly (i.e. `doseq` to issue some updates and return nil). So a lot of the work here is changing that to keep tabs on what's happened and return counts.

Below are the useful datapoints for sync
- [x] Average timings for each step of sync/analyze/field values
  - [x] Per database
  - [x] Globally
- [x] Sync/Analyze Summary
  - [x] Total number of fields sync'd
  - [x] Number of fields updated
  - [x] Number of query errors
  - [x] Number of fingerprints updated
  - [x] Number of fingerprints skipped
  - [x] Start/end time of sync/analysis
  - [x] Start/end of each stage in sync/analysis

senior commented on src/metabase/sync/sync_metadata/fields.clj:
```diff
@@ -16,10 +16,31 @@
             [schema.core :as s]
             [toucan.db :as db]))
 
+(deftimer um/reg sync-fields-timer)
```
I used timers, meters and histograms here. They can be def'd like I did here or identified by a string in the central registry.

Timers will give you counts and average timings for each invocation. Meters just count things as they go by and histograms will mark an event with a value.

senior commented on src/metabase/sync/sync_metadata/fields.clj:
```diff
@@ -264,14 +337,23 @@
-    (sync-fields-for-table! database table)))
+  (time! sync-and-update-fields-for-db-timer
+         (let [tables (sync-util/db->sync-tables database)]
+           (update! tables-per-db-hist (count tables))
```
What are the average number of tables per database sync'd

![tables-per-db](https://user-images.githubusercontent.com/215910/34177188-54ae4a12-e4c8-11e7-8304-e8f39ebe3e51.png)

senior commented on src/metabase/sync/sync_metadata/fields.clj:
```diff
@@ -264,14 +337,23 @@
    (sync-util/with-error-handling (format "Error syncing fields for %s" (sync-util/name-for-logging table))
      (let [db-metadata (db-metadata database table)]
+
+       (update! fields-per-table-hist (count db-metadata))
```
What are the average number of fields per table

![fields-per-table](https://user-images.githubusercontent.com/215910/34177219-771a774c-e4c8-11e7-97a1-da6cd61e4523.png)

senior commented on src/metabase/sync/sync_metadata/fields.clj:
```diff
@@ -264,14 +337,23 @@
+
        ;; make sure the instances of Field are in-sync
-       (sync-field-instances! table db-metadata (our-metadata table) nil)
+       (time! sync-fields-timer
```
On average, how long does it take us to sync fields?

![sync-fields-timer](https://user-images.githubusercontent.com/215910/34177229-81dbfc32-e4c8-11e7-8894-214200f96e85.png)

senior commented on src/metabase/sync/sync_metadata/fields.clj:
```diff
@@ -102,12 +143,14 @@
         {new-database-type :database-type, new-base-type :base-type} field-metadata]
     ;; If the driver is reporting a different `database-type` than what we have recorded in the DB, update it
     (when-not (= old-database-type new-database-type)
+      (mark! update-database-type-count)
```
Updating a database type seems like it should be a rare occurrence. Added this as it might be something of interest.

senior commented on src/metabase/sync/sync_metadata/fields.clj:
```diff
@@ -137,29 +183,45 @@
+              new-fields       (remove known-field-pred db-field-chunk)]
+
+          (update! existing-fields-hist (count fields-to-update))
+          (update! new-fields-hist      (count new-fields))
```
On average, how many fields are we updating vs. finding as new

![new-fields](https://user-images.githubusercontent.com/215910/34177324-dc98db68-e4c8-11e7-805f-71a99dd91923.png)

![existing-fields](https://user-images.githubusercontent.com/215910/34177328-e0e514d4-e4c8-11e7-9014-05bb12a4d564.png)

iethree committed:
Add timing and other metrics to sync/analyze/field values [ci drivers]

This commit adds a new "sync summary" message to sync/analyze/field
values scanning. It include information about how much data was
updated, begin/end timestamps around each phase etc.

senior commented:
Below is an example of the new output for sync and analyze
```
04-19 17:16:00 DEBUG sync.util :: Sync operations in flight: {:sync-metadata #{2}}                                                                                                                                                                                                [20/4739]
04-19 17:16:00 INFO sync.util :: STARTING: Sync metadata for mysql Database 2 'Magento'
04-19 17:16:00 DEBUG sync.util :: STARTING: step 'sync-timezone' for mysql Database 2 'Magento'
04-19 17:16:03 DEBUG sync.util :: FINISHED: step 'sync-timezone' for mysql Database 2 'Magento' (3 s)
04-19 17:16:03 DEBUG sync.util :: STARTING: step 'sync-tables' for mysql Database 2 'Magento'
04-19 17:16:03 DEBUG sync.util :: FINISHED: step 'sync-tables' for mysql Database 2 'Magento' (388 ms)
04-19 17:16:03 DEBUG sync.util :: STARTING: step 'sync-fields' for mysql Database 2 'Magento'
04-19 17:17:04 DEBUG sync.util :: FINISHED: step 'sync-fields' for mysql Database 2 'Magento' (1 mins)
04-19 17:17:04 DEBUG sync.util :: STARTING: step 'sync-fks' for mysql Database 2 'Magento'
04-19 17:17:24 DEBUG sync.util :: FINISHED: step 'sync-fks' for mysql Database 2 'Magento' (20 s)
04-19 17:17:24 DEBUG sync.util :: STARTING: step 'sync-metabase-metadata' for mysql Database 2 'Magento'
04-19 17:17:24 DEBUG sync.util :: FINISHED: step 'sync-metabase-metadata' for mysql Database 2 'Magento' (101 ms)
04-19 17:17:24 DEBUG sync.util ::
#################################################################
# Completed sync on Magento
# Start: 2018-04-19T22:16:00.095Z
# End: 2018-04-19T22:17:24.406Z
# Duration: 1 mins
# ---------------------------------------------------------------
# Completed step 'sync-timezone'
# Start: 2018-04-19T22:16:00.097Z
# End: 2018-04-19T22:16:03.187Z
# Duration: 3 s
# Found timezone id UTC
# ---------------------------------------------------------------
# Completed step 'sync-tables'
# Start: 2018-04-19T22:16:03.195Z
# End: 2018-04-19T22:16:03.584Z
# Duration: 389 ms
# Total number of tables sync'd 359, number of tables updated 0
# ---------------------------------------------------------------
# Completed step 'sync-fields'
# Start: 2018-04-19T22:16:03.585Z
# End: 2018-04-19T22:17:04.729Z
# Duration: 1 mins
# Total number of fields sync'd 3,432, number of fields updated 0
# ---------------------------------------------------------------
# Completed step 'sync-fks'
# Start: 2018-04-19T22:17:04.739Z
# End: 2018-04-19T22:17:24.299Z
# Duration: 20 s
# Total number of foreign keys sync'd 460, 0 updated and 0 tables failed to update
# ---------------------------------------------------------------
# Completed step 'sync-metabase-metadata'
# Start: 2018-04-19T22:17:24.303Z
# End: 2018-04-19T22:17:24.405Z
# Duration: 102 ms
#################################################################

04-19 17:17:24 INFO sync.util :: FINISHED: Sync metadata for mysql Database 2 'Magento' (1 mins)
04-19 17:17:24 DEBUG sync.util :: Sync operations in flight: {:analyze #{2}}
04-19 17:17:24 INFO sync.util :: STARTING: Analyze data for mysql Database 2 'Magento'
04-19 17:17:24 DEBUG sync.util :: STARTING: step 'fingerprint-fields' for mysql Database 2 'Magento'
04-19 17:17:24 INFO middleware.cache :: Using query processor cache backend: :db 💾
04-19 17:17:49 INFO sync.analyze :: fingerprint-fields Analyzed [*****·············································] 😢   10% Table 107 'dataflow_profile'
04-19 17:18:18 INFO sync.analyze :: fingerprint-fields Analyzed [**********········································] 😞   20% Table 259 'xmlconnect_notification_template'
04-19 17:18:45 INFO sync.analyze :: fingerprint-fields Analyzed [***************···································] 😕   30% Table 89 'sales_recurring_profile'
04-19 17:18:54 DEBUG sync.util :: FINISHED: step 'fingerprint-fields' for mysql Database 2 'Magento' (2 mins)
04-19 17:18:54 DEBUG sync.util :: STARTING: step 'calssify-fields' for mysql Database 2 'Magento'
04-19 17:18:55 INFO sync.analyze :: classify-fields Analyzed [********************······························] 😐   40% Table 146 'review_entity_summary'
04-19 17:18:55 INFO sync.analyze :: classify-fields Analyzed [*************************·························] 😬   50% Table 212 'paybox_question_number'
04-19 17:18:55 INFO sync.analyze :: classify-fields Analyzed [******************************····················] 😌   60% Table 339 'rating_store'
04-19 17:18:55 DEBUG sync.util :: FINISHED: step 'calssify-fields' for mysql Database 2 'Magento' (315 ms)
04-19 17:18:55 DEBUG sync.util :: STARTING: step 'classify-tables' for mysql Database 2 'Magento'
04-19 17:18:55 INFO sync.analyze :: clasify-tables Analyzed [***********************************···············] 😏   71% Table 56 'xmlconnect_config_data'
04-19 17:18:55 INFO sync.analyze :: clasify-tables Analyzed [****************************************··········] 😊   81% Table 221 'admin_rule'
04-19 17:18:55 INFO sync.analyze :: clasify-tables Analyzed [*********************************************·····] 😍   91% Table 331 'catalog_product_entity_datetime'
04-19 17:18:55 DEBUG sync.util :: FINISHED: step 'classify-tables' for mysql Database 2 'Magento' (173 ms)
04-19 17:18:55 DEBUG sync.util ::
#################################################################
# Completed analyze on Magento
# Start: 2018-04-19T22:17:24.450Z
# End: 2018-04-19T22:18:55.409Z
# Duration: 2 mins
# ---------------------------------------------------------------
# Completed step 'fingerprint-fields'
# Start: 2018-04-19T22:17:24.450Z
# End: 2018-04-19T22:18:54.916Z
# Duration: 2 mins
# Fingerprint updates attempted 471, updated 0, no data found 471, failed 0
# ---------------------------------------------------------------
# Completed step 'calssify-fields'
# Start: 2018-04-19T22:18:54.918Z
# End: 2018-04-19T22:18:55.235Z
# Duration: 317 ms
# Total number of fields classified 0, 0 failed
# ---------------------------------------------------------------
# Completed step 'classify-tables'
# Start: 2018-04-19T22:18:55.235Z
# End: 2018-04-19T22:18:55.408Z
# Duration: 173 ms
# Total number of tables classified 357, 0 updated
#################################################################

04-19 17:18:55 INFO sync.util :: FINISHED: Analyze data for mysql Database 2 'Magento' (2 mins)
```

camsaul commented on src/metabase/sync/util.clj:
```diff
@@ -260,3 +275,131 @@
+                                          :duration   (calculate-duration start-time end-time)
+                                          :steps      step-metadata})))
+
+(defn map+
```
wouldn't this be better called something like `sum`?

iethree committed:
Remove sync fiels min/max test [ci drivers]

This test predates the move of min/max values of a field to
fingerprints. Rmoving this test as it's covered over there and doesn't
provide a lot of value in the fields sync code.
